Direct rendering

erodozer opened the pull request:
~~currently builds on top of #124, leaving this as a draft PR until that is merged.~~ PR has been merged

Implements https://github.com/MizunagiKB/gd_cubism/issues/102

- GDCubismUserModel is now based on Node2D instead of Subviewport, allowing for direct rendering
- Mask shaders have been adjusted to work with this approach
- ~~mask_viewport_size as been changed to represent the shortest edge of the viewport texture, with its resolution matching the canvas aspect ratio to prevent visual errors~~ mask viewports are now dynamically sized based on mesh bounds to decrease memory requirements, users no longer need to manage mask scale

Adjust Pos and Adjust scale have been removed in favor of using position and scaling of Node2D.  As is standard practice in Godot, if you wish to adjust the offset of the model which is rendered from its TopLeft, simply wrap it with another Node2D.

erodozer commented:
~~I'm having big problems with VRAM usage on detailed models that I didn't notice before purely because I have a card with 16gb of VRAM :weary: Looking into a new way to do the mask shaders so the viewports don't have to be so large and details can still be maintained.~~

Fixed

erodozer commented:
New approach to masking has been implemented.

Mask subviewports are now created based on the bounding box of the mesh that is being masked, thus keeping sizes small.  This significantly improves VRAM consumption by drastically reducing the area needed to draw into.  Edge cases come from very large art meshes, though as long as the meshes are smaller than the total canvas size, it'll still be better than before.

Another small quirk with this approach is the viewport size can constantly change if the mesh that is masked is being animated.  Unlike before, the viewport will not be a fixed size, so VRAM consumption will not be constant after a model is loaded.  However, I feel the pros outweigh the cons here since the overall usage is lowered.

Before the adjustment with fixed sized viewports scaled to the canvas's aspect ratio
I couldn't even do canvas size on this model because I'd run out of VRAM and crash my session

With 2048x limit (mask matching canvas ratio)
![image](https://github.com/user-attachments/assets/705746bd-f526-495e-a6ae-ece96776fca6)

With dynamically sized mask viewports on an animated mesh we can see the fluctuation in memory usage.
![image](https://github.com/user-attachments/assets/2c174278-512a-42fc-ad8d-97ab8b318fe9)

JekSun97 commented:
I'm afraid that this will affect performance, since I don't have 16, but only 4 VRAM.

erodozer commented:
> I'm afraid that this will affect performance, since I don't have 16, but only 4 VRAM.

that's what the latest commit fixed.  From the screenshots I posted of the monitoring, I was able to drop the VRAM usage of a high resolution model from over 2gb of VRAM to about 300mb.

As a benchmark I was even able to load the model in both godot and vtube studio under similar conditions and could see that it was using equal or less VRAM than the unity live2d implementation.

JekSun97 commented:
Why am I worried? I'm currently drawing art for animation in Live2D, for a city simulator, where there will be many Live2D models as residents, and I'm wondering how many of these models Godot will be able to handle for display.

erodozer commented:
Heavy vram usage with Live2D stems from its textures, with masking causing additional textures to be allocated.  If you don't use high resolution models and limit your usage of masking you should be fine.

For reference, the model I've been testing with for my extreme case is the one that comes with vtube studio.  Vtuber live2d models tend to be high resolution because of the need to zoom in, as well as the advantage that they're usually the only object being rendered.  

This sample Live2D model has a canvas size of **5666x9004** as computed by the live2d SDK.  It has 95 generated Masks, and nearly 200 art meshes.  That canvas is way too massive to generate a masking viewport that covers the entire model just for drawing to a small area of it.  Which is why before this PR, the approach was to render the whole model in a subviewport and make masks relative to the subviewport size instead of the model's actual canvas size, unless overridden.

Here's some captured metrics of this model

**Before PR (aa16bac9998a8cb303b97ef508e55b5ca52a3856)**
Subviewport 512x1024 (~1/10th canvas scale), mask auto scale = true
![image](https://github.com/user-attachments/assets/1e85fc8e-7680-479b-9418-5dadfb8f6fa9)

Subviewport 2048x4096 (~1/2 canvas scale), mask auto scale = true
![image](https://github.com/user-attachments/assets/dc79f7a7-02a4-49a3-a8b3-d59463030848)

**With PR (cfc8b79bebc98f4f65cbde8cd6a2612caee5747f)**
direct render, dynamic masks
![image](https://github.com/user-attachments/assets/ca9877a9-84c5-4792-aac1-ade6deb581f3)

Direct Render w/ dynamic masks in this test case is able to show improved memory usage while rendering models at their full resolution, even compared to having to downscale things to about 1/10th of its original canvas.  Masks are now sharper and use less memory since they are sized to only the area they are affecting, and there's less complexity in the shaders.

MizunagiKB committed:
update document 0.8

MizunagiKB committed:
Document update

erodozer commented:
Something annoying to note in testing
https://github.com/godotengine/godot-proposals/issues/9428
https://github.com/godotengine/godot-proposals/issues/11038

Because of the z-index of model parts, overlapping models will have their layers interweave because sorting applies to the whole canvas.  Hopefully at some point in the near future one of these proposals will be implemented, but in the mean time the only workaround would be to wrap the model in a subviewport still.

erodozer commented:
I think I've found and squashed all the bugs I can in this approach, so I won't be making any more commits to this branch/PR unless there's anything needing addressed from review.

MizunagiKB committed:
Fixed type conversion error

MizunagiKB committed:
Merge pull request #133 from MizunagiKB/132-v080-gdcubismeffecthitarea-node-crash

Fixed type conversion error

Listare committed:
fix(GDCubismEffectTargetPoint): initialize `_need_update` to `true` to ensure `_map_param_idx` is set correctly

MizunagiKB committed:
Merge pull request #134 from Listare/fix/need-update-init

fix(GDCubismEffectTargetPoint): initialize `_need_update` to `true` to fix `_map_param_idx` initialization

JekSun97 commented:
I wanted to know what would happen if we wanted to embed the model in the user interface? as far as I know, Node2D nodes don't work very well with Control nodes, in terms of scale matrix, etc.

erodozer commented:
> I wanted to know what would happen if we wanted to embed the model in the user interface? as far as I know, Node2D nodes don't work very well with Control nodes, in terms of scale matrix, etc.

Node2D works fine within Control nodes with respect to their standard Canvas Item transform properties.  Scale, Position, and Rotation have consistent behavior between them, and even pivot offset works.  For reference, I'm using a Control wrapped Node2D Live2D model in my application so that I can drag it around with ease.  The only property that doesn't affect transformation of a Node2D is size, which describes the rectangle bounds of the control.

I also notice that a control's Clip Contents also doesn't affect Node2D children.  The shaders of the live2d models also don't respect the backbuffer manipulation used for Masking, so it doesn't work with CanvasItem visibiiltiy settings Clip Children of any parent node.

If you wish, you can still mimic the old behavior by explicitly creating your own subviewport and putting the live2d model in it.  However, as long as you do not need the specific functions that don't work properly without being in a subviewport, it's better to prefer direct rendering as it'll require less VRAM.

Also, since it doesn't respect Controls rectangle size, that means it won't autosize itself when used in layout containers.  This can be easily worked around with a script to adjust model scale and position when the control node is updated, which I'd recommend over using subviewport containers.

MizunagiKB committed:
Behavior Changes in NONE_PARAMETER Mode

MizunagiKB committed:
Changed to flags due to the limitations on player choices when using enums.

erodozer commented:
Was investigating bringing back the Mask Viewport Size feature, but as an upper bounds setting.  Through some testing, I was able to limit viewport sizes, while maintaining aspect ratio and scaling the content within.  The model I was testing has many small masks under 128x128, but some larger areas had masks over 2048x2048 in size.  By setting the upper bound, masks beyond the limit would be scaled down, and realistically I saw no visual degradation as I adjusted numbers from full resolution, 512x, even down to 32x

![image](https://github.com/user-attachments/assets/84fe3114-f3f7-42a5-bce8-fd3442f69ce8)

The impact is pretty promising on saving even more VRAM than has already been saved.

An ideal approach would be to combine upper bounds scaling with canvas item scale.  eg. if a model is scaled down and only taking up 128x128 px on the screen, there's no need for the viewport textures to be more than this.

These kinds of optimizations aren't present in the existing implementation, nor am I going to include it in this PR, as enough is already being done here.  Instead I'll be continuing investigation and work it into a follow-up optimization PR.

MizunagiKB committed:
The documentation has been updated.

MizunagiKB committed:
Updated C# script.

erodozer commented:
Definitely worth pursuing the automatic scaling optimizations.  By utilizing the viewport transformation of the model, I was able to detect if a mask will be visible and what its scaled resolution is going to be.  By applying those values, I could "hide" mask viewports entirely based on culling, or have them render much lower resolution with no perceivable impact on quality.  For high quality models that'll never be presented at 100% resolution, this has a huge benefit in reducing VRAM cost.

https://github.com/user-attachments/assets/1b972a12-2f9e-4385-8da6-c6afd1c77900

Live2D models with masks will be able to use less VRAM when scaled down and when masked parts are off screen.

MizunagiKB committed:
Merge pull request #135 from MizunagiKB/bugfix_none_parameter_mode

Bugfix none parameter mode

erodozer committed:
render live2d models directly without subviewport

- masks are dynamically sized based on art mesh dimensions and viewport scale
- mask viewport size acts as an upper bound for further control than what's automated
- removed properties no longer needed, such as scale and adjust position

erodozer commented:
I will be committing the changes for the mask resizing since it feels pretty critical to this for maximum performance and ease of use.

I am also squashing and rebasing the commits.

erodozer commented:
@MizunagiKB this should be good for review, and higher priority to merge than #128

MizunagiKB commented:
It has been quite some time, but I have finally reviewed this pull request.  
Regarding the functionality on macOS, I have confirmed that it works without any issues.

I also plan to adapt the _sample program_ while performing further testing.  

One aspect I noticed during testing is that the rendering position of the Live2D model is set to the upper-left corner. In Live2D, the origin is determined by the Export settings.

> Here is a relevant excerpt from the Live2D Cubism Core API Reference:

![fig1](https://github.com/user-attachments/assets/f12ec2d8-4093-41ef-bcb9-980b97892153)

While it is possible to calculate the origin or use a parent node to replicate the desired behavior, if the image is not rendered based on the position specified by the creator, it becomes cumbersome for Cubism Editor users to achieve their intended layout.

We would like the implementation to align with Live2D specifications.

The origin serves as the central point for scaling and rotation operations, and it becomes problematic when standardizing the origin at the feet of all characters to reflect height differences.

Although this may introduce some computational complexity, is it possible to implement rendering that reflects the Live2D specifications?  

> It is true that, at first glance, the implementation using SubViewport appears to render from the upper-left corner.
> However, this is not the intended behavior. The rendering performed within the SubViewport adheres to the Live2D specifications.

erodozer commented:
Oh, funny enough it appears that the cubism SDK already returns vertices relative to the origin.  I was expecting everything to be >= 0 but they actually go negative.  The adjustment that's being done of adding the origin to vertices was actually positioning everything to the top left.  Removing it will move the anchor to the origin

erodozer committed:
remove offset correction, vertices are already relative to canvas origin

MizunagiKB committed:
Port the sample program for Direct Rendering

MizunagiKB committed:
Port the sample program for Direct Rendering

MizunagiKB commented:
We have started by migrating the following three features. Other features will be migrated in due course:

- **viewer**, which is likely the first thing most users will run.
- **demo_effect_hit_area**, which utilizes mesh information.
- **demo_transparent**, which serves as a unique implementation example.

The major changes include the discontinuation of `SubViewport` usage and the shift of vertex calculations to the Shader side. During the migration, some points have drawn our attention, and we would like to share them here.

---

### Regarding Fade Processing

When using `SubViewport` previously, Live2D models were treated as a single texture, allowing for straightforward color calculations and fade processing as a `Sprite2D`.

With the current rendering approach, which processes parts individually, fade processing seems to require additional considerations. 

For users who wish to maintain similar functionality as before, we plan to add a single demo that compares SubViewport-based fading with PartOpacity-based fading.
However, most users are likely less concerned with `SubViewport` itself and more interested in achieving a seamless fade effect for their characters.

Do you have any recommendations for effective methods to address this?

---

### Regarding Rendering Priority

Since rendering is now performed on a per-part basis, the priority is determined individually based on the `z_index` of `GDCubismUserModel`. This change enables more versatile expressions, such as inserting `Sprite2D` elements between parts.

However, when multiple Live2D models are displayed together, conflicts in rendering order may occur. This issue can be resolved by setting `z_index` values incrementally, such as `1024` for the first `GDCubismUserModel`, `2048` for the second, and so on.

While not strictly necessary, it might be helpful to provide a mechanism to dynamically resolve this, such as the following:

```python
# Example:
var max_z = $GDCubismUserModel.get_max_z_index()
```

> That said, would this approach fail if the internal priority of the Live2D model is dynamically updated?

---

### Regarding Drag Operations(Editor)

Previously, since what users saw was a single texture, it was easy to enable drag functionality by attaching it to a `Sprite2D`. Currently, interaction is limited to the origin marker, which can be inconvenient.

By overriding `_edit_get_rect()` and `_edit_get_rect()` within `GDCubismUserModel` and utilizing the information used in `canvas_info`, it should be possible to implement drag operations.

Most users of `GDCubismUserModel` would likely share the same expectation.

If there is no particular reason for not enabling this, I am willing to implement it myself.  
> If it is not overly burdensome, you are welcome to handle the implementation instead.

MizunagiKB reviewed (dismissed):
While I will address the question above, the code itself and its practical use do not appear to have any issues.

There is a possibility that minor adjustments might be needed, but I suggest merging this into `main` for now.

Afterward, we can create a `0.9` branch to update the sample code and documentation and then merge them separately.

erodozer commented:
>Regarding Fade Processing

I was hoping [CanvasGroup](https://docs.godotengine.org/en/stable/classes/class_canvasgroup.html) could work, but since it relies on a custom shader and reading the backbuffer, we'd have to modify all the live2D shaders.  Reading the backbuffer could end up being more costly than subviewports as well at scale, since you want to minimize the number of shaders reading it.

The fade processing and z-index were both things I had noticed before, both of which I could not determine a suitable scalable solution for aside from using subviewports when the behavior is necessary.  

> Regarding Rendering Priority

The primary issue I have with the proposed solution is that dynamically resolving z-indexes can be complicated if you're not looking at the entire scene, as you would not want to introduce unexpected behavior when mixing live2d models with other types of 2d nodes like Sprites in a scene.

> Regarding Drag Operations(Editor)

I had not thought about the edit rect before, that would be nice to have.  Forgot that CanvasItems had that feature since I was too caught up on Node2D not having size rects like Controls do.  Feel free to add it

MizunagiKB committed:
Convert various demos for compatibility with Direct Rendering.

MizunagiKB commented:
I also tried using `CanvasGroup` and realized that modifying it on the shader side makes it exclusive to `CanvasGroup`. As for fading, I have added a demo for explanation purposes, which I believe will be helpful for those looking to implement fade processing.

> fade_demo

![fig3](https://github.com/user-attachments/assets/ca123349-ff41-4207-a447-7f47df21059a)

Regarding the priority of `z_index`, it is an issue that can be sufficiently resolved on the GDScript side, so let's leave it up to the users.

When I investigated how to implement drag operations, I found that `_edit_get_rect` and `_edit_set_rect` could only be used when embedding in the Editor.

To achieve this with `GDExtensions`, it seems necessary to create a class that inherits from `EditorPlugin`, register `GDCubismUserModel`, and customize the editing operations for each registered type.

Since `plugin.hpp` and `plugin.cpp` have already been added in another extension you're working on and are scheduled to be merged, I'll add this after the merge is complete.

MizunagiKB reviewed (approved):
The various demos have been updated to support Direct Rendering, and a new _fade_demo_ utilizing `SubViewport` has been added.

While the modifications for the **C#** version and the documentation updates are still in progress, it has been confirmed that the functionality in **GDScript** achieves comparable results to the previous implementation.
